From 5ff7cf20042f5bed6b898e2cda15712a755d91e5 Mon Sep 17 00:00:00 2001
From: Denis Kenzior <denkenz@gmail.com>
Date: Wed, 21 Oct 2020 16:55:47 -0500
Subject: uintset: Fix incomplete memcpy in _clone

l_uintset_clone was not calculating the bitmap size correctly and not
fully copying the bits of the original set.
---
 ell/uintset.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/ell/uintset.c b/ell/uintset.c
index d18ccfa..5a64d65 100644
--- a/ell/uintset.c
+++ b/ell/uintset.c
@@ -485,7 +485,8 @@ LIB_EXPORT struct l_uintset *l_uintset_clone(const struct l_uintset *original)
 	if (unlikely(!original))
 		return NULL;
 
-	bitmap_size = (original->size + BITS_PER_LONG - 1) / BITS_PER_LONG;
+	bitmap_size = sizeof(unsigned long) *
+		((original->size + BITS_PER_LONG - 1) / BITS_PER_LONG);
 
 	clone = l_uintset_new_from_range(original->min, original->max);
 	memcpy(clone->bits, original->bits, bitmap_size);
-- 
cgit 1.2.3-1.el7

